Speed up `Parser::expected_tokens`

nnethercote opened the pull request:
The constant pushing/clearing of `Parser::expected_tokens` during parsing is slow. This PR speeds it up greatly.

r? @estebank

nnethercote committed:
Rename `Parser::expected_tokens` as `Parser::expected_token_types`.

Because the `Token` type is similar to but different to the `TokenType`
type, and the difference is important, so we want to avoid confusion.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 01336018742501a6171b8c508cf9ebac7460d03f with merge 4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"01336018742501a6171b8c508cf9ebac7460d03f","merge_sha":"4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12136592591/job/33838044733) [(plain)](https://github.com/rust-lang/rust/commit/01336018742501a6171b8c508cf9ebac7460d03f/checks/33838044733/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking toml v0.7.8
error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/cfg_if.rs:34:36
     |
34   |             if !parser.eat_keyword(kw::If) {
     |                        ----------- ^^^^^^ expected `(Symbol, TokenType)`, found `Symbol`
     |                        arguments to this method are incorrect
     |
     = note: expected tuple `(rustc_span::Symbol, TokenType)`
               found struct `rustc_span::Symbol`
               found struct `rustc_span::Symbol`
note: method defined here
    --> /checkout/compiler/rustc_parse/src/parser/mod.rs:1195:12
     |
1195 |     pub fn eat_keyword(&mut self, (kw, token_type): (Symbol, TokenType)) -> bool {

error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/cfg_if.rs:58:24
     |
     |
58   |         if !parser.eat(&TokenKind::OpenDelim(Delimiter::Brace)) {
     |                    --- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `(&TokenKind, TokenType)`, found `&TokenKind`
     |                    arguments to this method are incorrect
     |
     |
     = note:  expected tuple `(&TokenKind, TokenType)`
             found reference `&TokenKind`
note: method defined here
     |
     |
1152 |     pub fn eat(&mut self, exp: (&TokenKind, TokenType)) -> bool {

error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/cfg_if.rs:81:24
     |
     |
81   |         if !parser.eat(&TokenKind::CloseDelim(Delimiter::Brace)) {
     |                    --- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `(&TokenKind, TokenType)`, found `&TokenKind`
     |                    arguments to this method are incorrect
     |
     |
     = note:  expected tuple `(&TokenKind, TokenType)`
             found reference `&TokenKind`
note: method defined here
     |
     |
1152 |     pub fn eat(&mut self, exp: (&TokenKind, TokenType)) -> bool {

error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/cfg_if.rs:85:23
     |
     |
85   |         if parser.eat(&TokenKind::Eof) {
     |                   --- ^^^^^^^^^^^^^^^ expected `(&TokenKind, TokenType)`, found `&TokenKind`
     |                   arguments to this method are incorrect
     |
     |
     = note:  expected tuple `(&TokenKind, TokenType)`
             found reference `&TokenKind`
note: method defined here
     |
     |
1152 |     pub fn eat(&mut self, exp: (&TokenKind, TokenType)) -> bool {

error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/cfg_if.rs:89:32
     |
     |
89   |         if !parser.eat_keyword(kw::Else) {
     |                    ----------- ^^^^^^^^ expected `(Symbol, TokenType)`, found `Symbol`
     |                    arguments to this method are incorrect
     |
     = note: expected tuple `(rustc_span::Symbol, TokenType)`
               found struct `rustc_span::Symbol`
               found struct `rustc_span::Symbol`
note: method defined here
    --> /checkout/compiler/rustc_parse/src/parser/mod.rs:1195:12
     |
1195 |     pub fn eat_keyword(&mut self, (kw, token_type): (Symbol, TokenType)) -> bool {

error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/lazy_static.rs:39:36
     |
     |
39   |         let _ = parser.eat_keyword(kw::Static);
     |                        ----------- ^^^^^^^^^^ expected `(Symbol, TokenType)`, found `Symbol`
     |                        arguments to this method are incorrect
     |
     = note: expected tuple `(rustc_span::Symbol, TokenType)`
               found struct `rustc_span::Symbol`
               found struct `rustc_span::Symbol`
note: method defined here
    --> /checkout/compiler/rustc_parse/src/parser/mod.rs:1195:12
     |
1195 |     pub fn eat_keyword(&mut self, (kw, token_type): (Symbol, TokenType)) -> bool {

error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/lazy_static.rs:40:36
     |
     |
40   |         let _ = parser.eat_keyword(kw::Ref);
     |                        ----------- ^^^^^^^ expected `(Symbol, TokenType)`, found `Symbol`
     |                        arguments to this method are incorrect
     |
     = note: expected tuple `(rustc_span::Symbol, TokenType)`
               found struct `rustc_span::Symbol`
               found struct `rustc_span::Symbol`
note: method defined here
    --> /checkout/compiler/rustc_parse/src/parser/mod.rs:1195:12
     |
1195 |     pub fn eat_keyword(&mut self, (kw, token_type): (Symbol, TokenType)) -> bool {

error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/lazy_static.rs:42:28
     |
     |
42   |         let _ = parser.eat(&TokenKind::Colon);
     |                        --- ^^^^^^^^^^^^^^^^^ expected `(&TokenKind, TokenType)`, found `&TokenKind`
     |                        arguments to this method are incorrect
     |
     |
     = note:  expected tuple `(&TokenKind, TokenType)`
             found reference `&TokenKind`
note: method defined here
     |
     |
1152 |     pub fn eat(&mut self, exp: (&TokenKind, TokenType)) -> bool {

error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/lazy_static.rs:44:28
     |
     |
44   |         let _ = parser.eat(&TokenKind::Eq);
     |                        --- ^^^^^^^^^^^^^^ expected `(&TokenKind, TokenType)`, found `&TokenKind`
     |                        arguments to this method are incorrect
     |
     |
     = note:  expected tuple `(&TokenKind, TokenType)`
             found reference `&TokenKind`
note: method defined here
     |
     |
1152 |     pub fn eat(&mut self, exp: (&TokenKind, TokenType)) -> bool {

error[E0308]: mismatched types
    --> src/tools/rustfmt/src/parse/macros/lazy_static.rs:46:28
     |
     |
46   |         let _ = parser.eat(&TokenKind::Semi);
     |                        --- ^^^^^^^^^^^^^^^^ expected `(&TokenKind, TokenType)`, found `&TokenKind`
     |                        arguments to this method are incorrect
     |
     |
     = note:  expected tuple `(&TokenKind, TokenType)`
             found reference `&TokenKind`
note: method defined here
     |
     |
1152 |     pub fn eat(&mut self, exp: (&TokenKind, TokenType)) -> bool {

error[E0308]: mismatched types
   --> src/tools/rustfmt/src/parse/parser.rs:110:36
    |
    |
110 |             match parser.parse_mod(&TokenKind::Eof) {
    |                          --------- ^^^^^^^^^^^^^^^ expected `(&TokenKind, TokenType)`, found `&TokenKind`
    |                          arguments to this method are incorrect
    |
    |
    = note:  expected tuple `(&TokenKind, TokenType)`
            found reference `&TokenKind`
note: method defined here
    |
57  |     pub fn parse_mod(
    |            ^^^^^^^^^

```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12136615452/job/33843653408)
Build commit: 4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4 (`4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12136615452/job/33843653408"},"merge_sha":"4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4"} -->

rust-timer commented:
Queued 4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4 with parent efdd9e802053caeb52103945df858e87f837e59a, future [comparison URL](https://perf.rust-lang.org/compare.html?start=efdd9e802053caeb52103945df858e87f837e59a&end=4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4): [comparison URL](https://perf.rust-lang.org/compare.html?start=efdd9e802053caeb52103945df858e87f837e59a&end=4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=efdd9e802053caeb52103945df858e87f837e59a&end=4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-2.4%, -0.2%] | 211   |
| Improvements ✅ <br /> (secondary) | -0.8% | [-2.6%, -0.1%] | 101   |
| All ❌✅ (primary)                 | -0.9% | [-2.4%, -0.2%] | 211   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=efdd9e802053caeb52103945df858e87f837e59a&end=4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4&stat=max-rss)
<details>
<summary>Results (primary -1.2%, secondary 1.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.5%, 0.5%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.9%  | [0.9%, 5.4%]   | 3     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-2.3%, -1.8%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.2%, -2.2%] | 1     |
| All ❌✅ (primary)                 | -1.2% | [-2.3%, 0.5%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=efdd9e802053caeb52103945df858e87f837e59a&end=4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4&stat=cycles:u)
<details>
<summary>Results (primary -1.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.5% | [-1.5%, -1.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.5% | [-1.5%, -1.5%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=efdd9e802053caeb52103945df858e87f837e59a&end=4e6952e2fa4367d9a5ef87505fa18f0dd3fedcc4&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 767.333s -> 766.554s (-0.10%)
**Artifact size**: 332.08 MiB -> 332.14 MiB (0.02%)

nnethercote committed:
Tweak some parser `check`/`eat` methods.

The most significant is `check_keyword`: it now only pushes to
`expected_token_types` if the keyword check fails, which matches how all
the other `check` methods work.

The remainder are just tweaks to make these methods more consistent with
each other.

nnethercote committed:
Remove `bra`/`ket` naming.

This is a naming convention used in a handful of spots in the parser for
delimiters. It confused me when I first saw it a long time ago, and I've
never liked it. A web search says "Bra-ket notation" exists in linear
algebra but the terminology has zero prior use in a programming context,
as far as I can tell.

This commit changes it to `open`/`close`, which is consistent with the
rest of the compiler.

nnethercote committed:
Speed up `Parser::expected_token_types`.

The parser pushes a `TokenType` to `Parser::expected_token_types` on
every call to the various `check`/`eat` methods, and clears it on every
call to `bump`. Some of those `TokenType` values are full tokens that
require cloning and dropping. This is a *lot* of work for something
that is only used in error messages and it accounts for a significant
fraction of parsing execution time.

This commit overhauls `TokenType` so that `Parser::expected_token_types`
can be implemented as a bitset. This requires changing `TokenType` to a
C-style parameterless enum, and adding `TokenTypeSet` which uses a
`u128` for the bits. (The new `TokenType` has 105 variants.)

The new types `ExpTokenPair` and `ExpKeywordPair` are now arguments to
the `check`/`eat` methods. This is for maximum speed. The elements in
the pairs are always statically known; e.g. a
`token::BinOp(token::Star)` is always paired with a `TokenType::Star`.
So we now compute `TokenType`s in advance and pass them in to
`check`/`eat` rather than the current approach of constructing them on
insertion into `expected_token_types`.

Values of these pair types can be produced by the new `exp!` macro,
which is used at every `check`/`eat` call site. The macro is for
convenience, allowing any pair to be generated from a single identifier.

The ident/keyword filtering in `expected_one_of_not_found` is no longer
necessary. It was there to account for some sloppiness in
`TokenKind`/`TokenType` comparisons.

The existing `TokenType` is moved to a new file `token_type.rs`, and all
its new infrastructure is added to that file. There is more boilerplate
code than I would like, but I can't see how to make it shorter.

nnethercote commented:
Best reviewed one commit at a time.

Let's re-run perf just to be sure: @bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit f5482df1eb2fac1c9a5bf73c1a4325bfff516d52 with merge 26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f5482df1eb2fac1c9a5bf73c1a4325bfff516d52","merge_sha":"26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b"} -->

estebank commented on compiler/rustc_parse/src/parser/diagnostics.rs:
```diff
@@ -2385,12 +2345,17 @@ impl<'a> Parser<'a> {
         Ok(param)
     }
 
-    pub(super) fn consume_block(&mut self, delim: Delimiter, consume_close: ConsumeClosingDelim) {
+    pub(super) fn consume_block(
+        &mut self,
+        open: ExpTokenPair<'_>,
+        close: ExpTokenPair<'_>,
+        consume_close: ConsumeClosingDelim,
+    ) {
```
Sad to see the delimiters go away and get denormalized, but I get it.

estebank commented on compiler/rustc_parse/src/parser/token_type.rs:
```diff
@@ -0,0 +1,616 @@
+/// `Parser::expected_token_types` was a `Vec<TokenType>` which was much slower
+/// to manipulate.
+///
+/// We really want to keep the number of variants to 128 or fewer, sot that
+/// `TokenTypeSet` can be implemented with a `u128`.
```
On the one hand, we should be able to do so. On the other, I can see this becoming a point of contention with t-lang in the medium future if we push back on a feature for this reason :)

nnethercote replied:
The 17 asm symbols would be a good place to cut things down if necessary, I'm a bit annoyed that they are even in there; so many of them for such a rare use case.

estebank commented on compiler/rustc_parse/src/parser/token_type.rs:
```diff
@@ -0,0 +1,616 @@
+
+            _ => panic!("unhandled value: {val}"),
+        };
+        // This assertion will detect if this method and the type definition get out of sync.
+        assert_eq!(token_type as u32, val);
+        token_type
```
Can't the function just be the `as` cast with a `<=104` check?

nnethercote replied:
No. You can convert a C-style enum to an integer with `as`, but you can't convert in the other direction, e.g. as per [this StackOverflow answer](https://stackoverflow.com/questions/28028854/how-do-i-match-enum-values-with-an-integer/29530566). There are proc macros to do it, but that answer pointed out an alternative that is suitable here: `transmute` is fine so long as the enum is `repr(uN)` for some value of `N`. So I will do that with `repr(u8)`, which will cut over 100 lines of code, yay.

estebank replied:
Should we have a static assertion that all of them roundtrip? I'm always concerned about a careless future reformat breaking the bidirectional mapping.

nnethercote replied:
I'd be happy for extra protection, but I'm having trouble imagining what such a static assertion would look like. Can you explain more?

nnethercote replied:
Alternatively, I can go back to an explicit match. The StackOverflow answer mentioned this style:
```
match v {
            x if x == MyEnum::A as i32 => Ok(MyEnum::A),
            x if x == MyEnum::B as i32 => Ok(MyEnum::B),
            x if x == MyEnum::C as i32 => Ok(MyEnum::C),
            _ => Err(()),
        }
```
It requires a line for every variant, but avoids having to write a number on each line.

estebank replied:
> I'd be happy for extra protection, but I'm having trouble imagining what such a static assertion would look like. Can you explain more?

I'd forgotten that the range operation doesn't work today in const, but I was picturing something like:

```rust
const __CHECK: () = const {
    for i in 0..2 {
        assert_eq!(E::to_i32(&E::from_i32(i)), i);
    }
};
```

nnethercote replied:
I've gone with the guard-based version, using a macro to avoid excessive boilerplate. It doesn't rely on `unsafe`, and also doesn't rely on matching up the right integer with the right variant.

estebank commented on compiler/rustc_parse/src/parser/token_type.rs:
```diff
@@ -0,0 +1,616 @@
+// bogus "macro-expanded `macro_export` macros from the current crate cannot be
+// referred to by absolute paths" error, ugh. See #52234.
+#[cfg_attr(rustfmt, rustfmt::skip)]
+macro_rules! exp {
```
Just to make sure I understand, `exp` and `Exp*` stands for "expected"?

nnethercote replied:
Yes.

estebank commented on compiler/rustc_parse/src/parser/token_type.rs:
```diff
@@ -0,0 +1,616 @@
+    (Question)       => { exp!(@tok, Question,  Question) };
+    (Eof)            => { exp!(@tok, Eof,       Eof) };
+
+    (Plus)           => { exp!(@binop, Plus) };
+    (Minus)          => { exp!(@binop, Minus) };
+    (Star)           => { exp!(@binop, Star) };
+    (And)            => { exp!(@binop, And) };
+    (Or)             => { exp!(@binop, Or) };
```
I'm half tempted at making this support things like `exp!(+)` and `exp!(?)` 😈

nnethercote replied:
I was initially going to do that. At one point I had `exp!(_)` instead of `exp!(Underscore)` and I was using actual keywords like `exp!(if)` instead of `exp!(If)`. A lot of tokens can be made to work but all the delimiters absolutely cannot, so I decided to use `StudlyCaps` names for all of them :)

estebank commented:
I'll finish reviewing tomorrow

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12154091930/job/33896752298)
Build commit: 26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b (`26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12154091930/job/33896752298"},"merge_sha":"26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b"} -->

rust-timer commented:
Finished benchmarking commit (26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b): [comparison URL](https://perf.rust-lang.org/compare.html?start=3b382642aba7cffbb2f47829b24635fad87bcf5c&end=26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=3b382642aba7cffbb2f47829b24635fad87bcf5c&end=26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-2.4%, -0.2%] | 213   |
| Improvements ✅ <br /> (secondary) | -0.8% | [-2.7%, -0.1%] | 98    |
| All ❌✅ (primary)                 | -0.9% | [-2.4%, -0.2%] | 213   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3b382642aba7cffbb2f47829b24635fad87bcf5c&end=26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b&stat=max-rss)
<details>
<summary>Results (primary -0.9%, secondary 0.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.4%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 1.7%  | [0.7%, 2.3%]   | 3     |
| Improvements ✅ <br /> (primary)   | -2.2% | [-2.2%, -2.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.8% | [-1.8%, -1.8%] | 1     |
| All ❌✅ (primary)                 | -0.9% | [-2.2%, 0.4%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3b382642aba7cffbb2f47829b24635fad87bcf5c&end=26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b&stat=cycles:u)
<details>
<summary>Results (primary 1.4%, secondary 2.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.4% | [0.9%, 2.5%] | 22    |
| Regressions ❌ <br /> (secondary)  | 2.4% | [1.3%, 3.2%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.4% | [0.9%, 2.5%] | 22    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=3b382642aba7cffbb2f47829b24635fad87bcf5c&end=26060e63f06a4dcd55fc0757eb5b0bdc8136ed3b&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 767.635s -> 770.294s (0.35%)
**Artifact size**: 330.89 MiB -> 330.89 MiB (0.00%)

nnethercote commented:
I updated, adding a new commit that uses `transmute` for the int-to-`TokenType` conversion.

nnethercote committed:
Make `TokenType::from_u32` foolproof.

Currently it relies on having the right integer for every variant, and
if you add a variant you need to adjust the integers for all subsequent
variants, which is a pain.

This commit introduces a match guard formulation that takes advantage of
the enum-to-integer conversion to avoid specifying the integer for each
variant. And it does this via a macro to avoid lots of boilerplate.

bors commented:
:umbrella: The latest upstream changes (presumably #134039) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased.

@estebank, are you happy with the updated final commit?

bors commented:
:umbrella: The latest upstream changes (presumably #129514) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased again.

nnethercote commented:
I rebased a third time.

bors commented:
:umbrella: The latest upstream changes (presumably #134243) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased a fourth time.

bors commented:
:umbrella: The latest upstream changes (presumably #134470) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
r? @spastorino

Santiago, initial review was done by Esteban two weeks ago, who we haven't heard from since. There was some back-and-forth about the `TokenType::from_u32` function, and the final commit puts it into what I think is good shape. Individual commit logs have explanations of what changes are made, and I'm happy to answer questions. Thanks!

spastorino commented:
r=me with CI green

nnethercote commented:
CI is green.

@bors r=spastorino

bors commented:
:pushpin: Commit 5ecae1b54d69374dc603d3e8c8304c3871bf53cc has been approved by `spastorino`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=spastorino 5ecae1b54d69374dc603d3e8c8304c3871bf53cc -->
<!-- homu: {"type":"Approved","sha":"5ecae1b54d69374dc603d3e8c8304c3871bf53cc","approver":"spastorino","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 5ecae1b54d69374dc603d3e8c8304c3871bf53cc with merge e46b2c453cbcb2f95ccc20904d6944711c1c9aa4...
<!-- homu: {"type":"BuildStarted","head_sha":"5ecae1b54d69374dc603d3e8c8304c3871bf53cc","merge_sha":"e46b2c453cbcb2f95ccc20904d6944711c1c9aa4"} -->

rust-log-analyzer commented:
The job **`dist-s390x-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12404782297/job/34630516759) [(plain)](https://github.com/rust-lang-ci/rust/commit/e46b2c453cbcb2f95ccc20904d6944711c1c9aa4/checks/34630516759/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
error[E0308]: mismatched types
   --> /rustc/e46b2c453cbcb2f95ccc20904d6944711c1c9aa4/compiler/rustc_index/src/lib.rs:40:32
    |
38  | macro_rules! static_assert_size {
    | ------------------------------- in this expansion of `rustc_data_structures::static_assert_size!`
39  |     ($ty:ty, $size:expr) => {
40  |         const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a size of 288, found one with a size of 280
   ::: compiler/rustc_parse/src/parser/mod.rs:194:56
    |
    |
194 | rustc_data_structures::static_assert_size!(Parser<'_>, 288);
    | |                                                      |
    | |                                                      help: consider specifying the actual array length: `280`
    | in this macro invocation

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12404782297/job/34631494835)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12404782297/job/34631494835","builder_name":"checks-actions"} -->

jieyouxu commented:
Array len mismatch.
@rustbot author

nnethercote committed:
Fix `Parser` size assertion on s390x.

For some reason the memory layout is different on s390x.

nnethercote commented:
s390s size assertion failure, fun.

@bors try

bors commented:
:hourglass: Trying commit 0f7dccf784dd46156281dff11f0be31d43f9bbcd with merge 7ebe629a8d2d74c9bf12e339350633096488d6ae...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0f7dccf784dd46156281dff11f0be31d43f9bbcd","merge_sha":"7ebe629a8d2d74c9bf12e339350633096488d6ae"} -->

jieyouxu commented on compiler/rustc_parse/src/parser/token_type.rs:
```diff
@@ -0,0 +1,631 @@
+use rustc_ast::token::TokenKind;
+use rustc_span::symbol::{Symbol, kw, sym};
+
+/// Used in "expected"/"expected one of" error messages. Tokens are added here
+/// as necessary. Tokens with values (e.g. literals, identifiers) are
+/// represented by a single variant (e.g. `Literal`, `Ident`).
+///
+/// It's an awkward representation, but it's important for performance. It's a
+/// C-style parameterless enum so that `TokenTypeSet` can be a bitset. This is
+/// important because `Parser::expected_token_types` is very hot. `TokenType`
+/// used to have variants with parameters (e.g. all the keywords were in a
+/// single `Keyword` variant with a `Symbol` parameter) and
+/// `Parser::expected_token_types` was a `Vec<TokenType>` which was much slower
+/// to manipulate.
+///
+/// We really want to keep the number of variants to 128 or fewer, so that
+/// `TokenTypeSet` can be implemented with a `u128`.
```
Suggestion: @nnethercote could you also include this comment (or even better, a short high-level description of this scheme) in the PR description, if only to make it easier to search a few years down the line (i.e. when someone wants to look up "huh, why is the `TokenType` representation in this form")?

nnethercote replied:
Um, isn't a comment on `TokenType` better for answering that question than a description in the PR that modified it?

jieyouxu replied:
That's fine too.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12409582851/job/34648587329)
Build commit: 7ebe629a8d2d74c9bf12e339350633096488d6ae (`7ebe629a8d2d74c9bf12e339350633096488d6ae`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12409582851/job/34648587329"},"merge_sha":"7ebe629a8d2d74c9bf12e339350633096488d6ae"} -->

bors commented:
:pushpin: Commit 0f7dccf784dd46156281dff11f0be31d43f9bbcd has been approved by `spastorino`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=spastorino 0f7dccf784dd46156281dff11f0be31d43f9bbcd -->
<!-- homu: {"type":"Approved","sha":"0f7dccf784dd46156281dff11f0be31d43f9bbcd","approver":"spastorino","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
Some try jobs are not getting picked up correctly, trying a sync cc https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/try.20jobs.20not.20kicking.20off.
@bors treeclosed=1000

jieyouxu commented:
@bors r- retry (sync)

nnethercote commented:
The tree seems to be back open?

@bors r=spastorino

bors commented:
:pushpin: Commit 0f7dccf784dd46156281dff11f0be31d43f9bbcd has been approved by `spastorino`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=spastorino 0f7dccf784dd46156281dff11f0be31d43f9bbcd -->
<!-- homu: {"type":"Approved","sha":"0f7dccf784dd46156281dff11f0be31d43f9bbcd","approver":"spastorino","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 0f7dccf784dd46156281dff11f0be31d43f9bbcd with merge 9e136a30a965bf4e63f03095c57df7257bf96fd6...
<!-- homu: {"type":"BuildStarted","head_sha":"0f7dccf784dd46156281dff11f0be31d43f9bbcd","merge_sha":"9e136a30a965bf4e63f03095c57df7257bf96fd6"} -->

jieyouxu commented:
Oops yeah, sorry about that -- bors picked up treeclosed from *another* PR, and I lost track of *this* PR due to notifications flood...

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12419866547/job/34682595404)
Approved by: spastorino
Pushing 9e136a30a965bf4e63f03095c57df7257bf96fd6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"spastorino","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12419866547/job/34682595404"},"merge_sha":"9e136a30a965bf4e63f03095c57df7257bf96fd6"} -->

rust-timer commented:
Finished benchmarking commit (9e136a30a965bf4e63f03095c57df7257bf96fd6): [comparison URL](https://perf.rust-lang.org/compare.html?start=11663cd3bfefef7d34e8f0892c250bf698049392&end=9e136a30a965bf4e63f03095c57df7257bf96fd6&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=11663cd3bfefef7d34e8f0892c250bf698049392&end=9e136a30a965bf4e63f03095c57df7257bf96fd6&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-2.5%, -0.2%] | 213   |
| Improvements ✅ <br /> (secondary) | -0.8% | [-2.5%, -0.1%] | 105   |
| All ❌✅ (primary)                 | -0.9% | [-2.5%, -0.2%] | 213   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=11663cd3bfefef7d34e8f0892c250bf698049392&end=9e136a30a965bf4e63f03095c57df7257bf96fd6&stat=max-rss)
<details>
<summary>Results (primary -0.1%, secondary -0.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.0%  | [2.0%, 2.1%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 4.8%  | [4.8%, 4.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.5% | [-2.1%, -0.9%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-3.7%, -1.8%] | 4     |
| All ❌✅ (primary)                 | -0.1% | [-2.1%, 2.1%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=11663cd3bfefef7d34e8f0892c250bf698049392&end=9e136a30a965bf4e63f03095c57df7257bf96fd6&stat=cycles:u)
<details>
<summary>Results (primary -1.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-1.7%, -1.0%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.2% | [-1.7%, -1.0%] | 4     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=11663cd3bfefef7d34e8f0892c250bf698049392&end=9e136a30a965bf4e63f03095c57df7257bf96fd6&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 769.451s -> 766.727s (-0.35%)
**Artifact size**: 330.39 MiB -> 330.36 MiB (-0.01%)

cuviper commented:
> The job **`dist-s390x-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12404782297/job/34630516759) [(plain)](https://github.com/rust-lang-ci/rust/commit/e46b2c453cbcb2f95ccc20904d6944711c1c9aa4/checks/34630516759/logs)
> Click to see the possible cause of the failure (guessed by this bot)
> 
> ```
>    Compiling rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
> error[E0308]: mismatched types
>    --> /rustc/e46b2c453cbcb2f95ccc20904d6944711c1c9aa4/compiler/rustc_index/src/lib.rs:40:32
>     |
> 38  | macro_rules! static_assert_size {
>     | ------------------------------- in this expansion of `rustc_data_structures::static_assert_size!`
> 39  |     ($ty:ty, $size:expr) => {
> 40  |         const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
>     |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a size of 288, found one with a size of 280
>    ::: compiler/rustc_parse/src/parser/mod.rs:194:56
>     |
>     |
> 194 | rustc_data_structures::static_assert_size!(Parser<'_>, 288);
>     | |                                                      |
>     | |                                                      help: consider specifying the actual array length: `280`
>     | in this macro invocation
> ```

> s390s size assertion failure, fun.

I just got the same error with 280 on powerpc64le in a Fedora scratch build of 1.85.0-beta.5:
https://kojipkgs.fedoraproject.org//work/tasks/4794/128284794/build.log

I have no idea why it would be different than Rust CI though, which apparently did pass on that arch...

cuviper commented:
I think the difference is due to the alignment of `TokenTypeSet(u128)`, which is only 8 on ppc64 and s390x... at least until #134115 takes effect on ppc64 with LLVM 20. But again, I don't know why Rust CI didn't see that too!

cuviper commented:
> I have no idea why it would be different than Rust CI though

I figured it out -- #134115 changed ppc64 alignment, but when I'm building natively it's starting with the old alignment in stage0, while Rust CI only cross-compiles using the last stage.
